### ISSUE TITLE
Fixing far apart siblings issue

### ISSUE THREAD

**Issue by nhadlaw** [touretzkyds/KnowledgeGraphDemo#44]
So I've been doing a lot of research and experimenting with the knowledge graph demo to try and fix this issue where some siblings are being displayed far apart and I have a couple of comments/potential fixes. So if we look at an example where this is occurring (Eptesicus):
![image 1](https://user-images.githubusercontent.com/101290675/176285949-32d773e3-d9a5-4d6f-a0f7-58f349294b1c.png)
I believe the issue is with the Cytoscape ranking algorithm because if I look at the other few instances of this issue, the parent node of these siblings is always a genus that is at the same level as a lot of species nodes from a different parent node beside them. If I look at the other genus parent nodes that this issue doesn't occur in, the other parent nodes beside them are genus nodes that are at the same level. 

I experimented with my algorithm in demo2d.html to see if I could try to force the dagre algorithm to display nodes in a certain order but that also had no effect. I tried creating the nodes in alphabetical order, in non-alphabetical order,  in parent and all children order, and I tried changing the node names to being just the qvalues (so they were all completely unique).All of this led to the same result where the nodes that were so far apart were still being displayed in the exact same spot. So my next approach was to see if altering the depth of these genus nodes would help solve the ranking issue. At first I tried adding a temporary node in between a genus node with the issue (like Eptesicus) and it's parent. That led to the same result:
![image 2](https://user-images.githubusercontent.com/101290675/176289113-159742b5-d340-4237-97ab-58568e568137.png)

So then I tried taking away the previous parent node of the genus node and that seemed to fix the issue:
![image 3](https://user-images.githubusercontent.com/101290675/176289234-7a08c838-a0c4-46d0-9efa-0f8104b8df5c.png)

I.e. notice now how the two children of Eptesicus are no longer absurdly far apart but I had to remove the original parent node of Eptesicus which was Eptesicini of rank "Tribe".  

I know this fix is not ideal but I think it's good to know that the issue is coming from the issue of inconsistent "Genus" depths in the tree (Specifically genus nodes being at the same level as species nodes). I could try manually fixing the positioning of these few (less than 5 genus parent nodes ) instances where some siblings are far apart but that would also not be ideal. I was wondering what you suggest I should try next? Again it is only occurring with a few number of nodes so whatever change we make wouldn't affect much of the graph at all. Another thing I was thinking of trying would be to take away nodes above the genus nodes that the issue is occurring at and then later add them back in? I will continue to look through the dagre documentation/code to see if I can alter anything but it doesn't look too promising.

Let me know what you think is best.

**Comment by touretzkyds** [comment]
I haven't studied the dagre algorithm so I don't fully understand what you're telling me. Are you assigning depths to nodes based on their taxon rank, so that cytoscape expects all genus nodes should be at the same depth? And when they're not, cytoscape does something weird like we're seeing?

If this is the case, wouldn't a solution be to simply assign depths to nodes based on their distance from the root? Why should we assume that all genus nodes should be at the same depth?

**Comment by nhadlaw** [comment]
No, I'm not assigning ranks for the nodes. I think the dagre algorithm is implicitly assigning ranks (not taxon ranks just numerical ranking) to nodes based on their depth from the root. (By the way, the same issues occur with the Full display so the mode of display doesn't have to do with the issue). 

I don't know if this example will help but if let's look at an instance where this issue is happening. Once again, Eptesicus is a Genus node that displays its two children far apart. In the Concise display mode, 
- The depth for Eptesicus (rank: Genus) from Animalia is 16. 
-  The depth for Alouatta arctoidea boltz: Q1207222 (rank: Species) is 16. This is the closest node (at the same level) to the right of the Eptesicus node.
- Depth for a node Aotus (rank: Genus), where this issue doesn't occur,  is 15.
Here is a visual of what I'm saying:
![image demo](https://user-images.githubusercontent.com/101290675/176508789-6ac8227e-3dd4-4757-ad15-bc8cb00770b0.png)
The point of this example is to show that the dagre algorithm that cytosape uses is confusing some nodes, which are at the 
same depth from Animalia, as siblings when they're not supposed to be. So as a result of this it's messing up the grouping of the children of these mistaken siblings. This is what makes the most sense to me as of what's going on. It's not a perfect answer but I think my next step is to continue to look through the dagre algorithm and see if there's anything I can change in terms of how it ranks the nodes but I don't know how helpful that will be.

**Comment by touretzkyds** [comment]
What you're showing me in the figure above does not make sense. You drew blue arrows to two children of Eptesicus and said they were species, but they can't be, because they have nodes below them. I pulled the latest version of the devel branch, and what I see in demo2d is that Eptesicus is a genus with parent Eptesicini (a tribe) and two children which are species. There is nothing below them. And in reality these two species are children of subgenus nodes, not the genus node.

So switch from Concise to Full view, and now what we see is a mess:
![bad-nodes](https://user-images.githubusercontent.com/1978092/176521075-e16ffefe-fab3-4a0c-9526-2088bf221842.png)

If I look in Fuseki it appears that Eptesicus-genus has two children, Eptesicus-subgenus and Cnephaeus-subgenus. Why does Eptesicus-genus have two separate subTaxonOf links to Eptescini-tribe? It looks like you're still building the graph wrong.

**Comment by nhadlaw** [comment]
Yes, it was my bad saying the children of Eptesicus were Species but that wasn't really important in my example. The key thing to see is that I think some genus nodes (like Eptesicus) and a species node (children of Alouatta) are being mistaken as siblings because of they have the same depth from the root node. Also the issue you showed in the picture is occurring most likely because when a user enters a taxon in the text box, they type in the taxonName not the Q value. This implies that my algorithm is getting confused by the fact that there are two nodes with the name Eptesicus. I can fix this issue pretty easily by either prompting another textbox that asks for the rank of the node if it's a species that shares the same exact name with another node or just automatically picking the Genus version of the node.

**Comment by touretzkyds** [comment]
I don't understand why two nodes having the same depth from the root should cause them to be mistaken as siblings. You think this is the explanation for what's going on with dagre, but I'm not convinced, since it only happens occasionally. Why doesn't it happen all the time? And Eptesicus is a problematic example because of the complicating factor that there are two nodes with the same name. It's suspicious. Show me an example where we don't have this complicating factor.

As for what to do with the text box input: if there are two nodes with the same name, choose the node with the higher taxon rank. So for Eptesicus you should choose the genus node over the subgenus node, but don't hard-code "genus" into this. Do it based on relative taxon rank.

**Comment by nhadlaw** [comment]
Looking at the full display mode, here are a good portion of instances of this issue happening:
1.  Panthera:
![image 1](https://user-images.githubusercontent.com/101290675/176553339-b24f106a-e88e-4e82-9b78-8146877dd19b.png)
2. Eptesicus (which is the above ex)
3. Alouatta (in red) and Lagothrix (in blue)
![image 2](https://user-images.githubusercontent.com/101290675/176553714-f6f1f1fe-b74f-4e4f-91a2-66514e9af833.png)
4. Cebus (in blue), Cercopithecini (in red)
![image 3](https://user-images.githubusercontent.com/101290675/176554355-574c82af-49fb-47e6-ab8d-94a9375dff4d.png)
5. Pongo
![image 4](https://user-images.githubusercontent.com/101290675/176554463-fd177ccd-b332-4295-952a-7b2bc4357780.png)
6. Pithecia
![image 5](https://user-images.githubusercontent.com/101290675/176554607-fa99ca5b-87e6-43d7-84bf-e84a16c4e782.png)

I tried to find a common pattern among all of the instances where this issue was occurring and the strongest connection between all of them is that the parent is a node that is at the same depth as a species node beside it. I don't know if this has anything to do with the issue but that is something I found in almost all of the instances where this issue occurred.

**Comment by nhadlaw** [comment]
Ok I think I found something that could have potential for fixing this issue. I have uploaded a file called demo2dfix.html just for the purposes of showing you what the fix makes the graph look like (I can remove this file once we decide on it). If you open this file, the Full graph loads (which is the ONLY display that should work, again this is just for demostration). The only thing I did to change the graph was added an extra child (I just called it temp node) to all species nodes (the nodes in black). The issue seems to have been pretty much fixed. I think this confirms the depth issue I was talking about with some species vs. other rank nodes.

**Comment by nhadlaw** [comment]
One last note but if we wanted to go with this potential fix, I could just make these added node sizes 0 and the graph doesn't look like it has extra nodes but it also won't have this sibling issue anymore.

**Comment by touretzkyds** [comment]
I'm just posting this Eptesicus query here to preserve it, in case we want to use it for further debugging later.
````
PREFIX kgo: <http://solid.boltz.cs.cmu.edu:3030/ontology/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

SELECT ?x ?rLabel ?y ?yLabel WHERE {
  ?x rdfs:label|skos:prefLabel "Eptesicus"@en.
  ?x ?r ?y.
  #?y ?r ?x.
  OPTIONAL {?r rdfs:label|skos:prefLabel ?rLabel}
  OPTIONAL {?y rdfs:label|skos:prefLabel ?yLabel}
 } ORDER BY ?x ?r ?y
````

**Comment by touretzkyds** [comment]
I don't think demo2dfix has fixed the issue. Look at the weird layout for Eptesicus:

![Eptesicus](https://user-images.githubusercontent.com/1978092/176582074-790862e7-c048-4644-938a-e21a1d897345.png)

**Comment by touretzkyds** [comment]
We have to look more deeply into how the dagre layout algorithm works. I think it has two conflicting goals and it uses some kind of heuristic to choose between them. Let's look at a small subgraph centered on Lepidoptera:

![Lepidoptera](https://user-images.githubusercontent.com/1978092/176594503-8174abc9-602c-4079-9214-c763cf3e3c7e.png)

Looking at just the bottom layers, we see three evenly spaced columns. This is the first goal: to space chiildren evenly, resulting in a regular, grid-like appearance. Even though Eurytelini and Satyrini are siblings while Melipotini is their cousin, all three are at the same depth and are evenly spaced.

The second goal is to center a parent over its children. Look at Nymphalidae. It is evenly spaced between Eurytelini and Satyrini, even though this position does not align with any of the three columns. The two goals are in conflict, and in the case of Nymphalidae the second goal won and we've violated the columnar layout.

Now look at Lepidoptera. It is not evenly spaced between Erebidae and Nymphalidae, as goal 2 requires. Instead it is lined up with the middle column; it's directly above Eurytelini. So for Lepidoptera the first goal won out over the second goal, and we end up with an asymmetry with respect to its children: the link from Erebidae is longer than the link from Nymphalidae. This looks ugly if you zoom in on just those nodes, but if you zoom all the way out like in the figure above, the overall structure is actually pretty nice: it has three clear columns and only one node (Nymphalidae) that doesn't line up with those columns. So if you really like grids, this is a great result. If you are a stickler for symmetry, then you might prefer to always give priority to the second goal, even though you lose the nice grid look. In that case dagre might not be the best algorithm to use.

When a node has an even number of children we cannot satisfy both goals at the same time. The most dramatic case is with just two children. Dagre seems to favor symmetry, at least when all the children fit into just two columns as with Nymphalidae. But it doesn't always do so even in this special case. Consider the layout for Equus-genus, which is forced into alignment with the first child Equus-subgenus, leaving the second child, Hippotigris, looking very asymmetric:

![Equus](https://user-images.githubusercontent.com/1978092/176598286-cb1bb8bf-ccc7-4bdb-84ec-8dc370bfdd31.png)

Dagre is using some kind of heuristic to trade off between the two goals. We don't know what that heuristic is; you'll have to read the source code and puzzle it out. I'm guessing that the heuristic has some limitations, such as perhaps only considering a limited depth subgraph when laying out a node, and this is resulting in bad decisions in cases where the graph is very wide, very deep, and very irregular, like our taxonomy. That's why we're seeing these weird layout decisions.

**Comment by touretzkyds** [comment]
I've been looking more closely at dagre. It's old code, deprecated and no longer maintained. It's built on top of something called lodash which is really old and really deprecated.

I think we should consider using something else. For simple tree layout it's easy to write a recursive algorithm that generates tight trees with symmetric parent placement. I can show you the algorithm. I'm not sure how easy it would be to get the columnar organization that dagre produces, but I don't think that's important.

For more complex graph layouts I suggest we look at cola.js. It's a more modern package and is fully compatible with Cytoscape. The docs say it's not intended for large graphs, but we can use our recursive tree layout algorithm for that. The web site is here: **https://ialab.it.monash.edu/webcola/**
